src/tree_data_sorted.c

```diff
@@ -0,0 +1,958 @@
+ * @return 1 if node is alone otherwise 0.
+ */
+#define DNODE_IS_ALONE(NODE) \
+    ((NODE->prev == NODE) || \
+        ((NODE->prev->schema != NODE->schema) && (!NODE->next || (NODE->schema != NODE->next->schema))))
+
+/**
+ * @brief Get Red-black tree from data node.
+ *
+ * @param[in] leader First instance of the (leaf-)list in sequence.
+ * @param[out] meta Metadata from which the Red-black tree was obtained. The parameter is optional.
+ * @return Root of the Red-black tree or NULL.
+ */
+static struct rb_node *
+lyds_get_rb_tree(const struct lyd_node *leader, struct lyd_meta **meta)
+{
+    struct rb_node *rbt;
+    struct lyd_meta *iter;
+
+    if (meta) {
+        *meta = NULL;
+    }
+    LY_LIST_FOR(leader->meta, iter) {
+        if (!strcmp(iter->name, RB_NAME)) {
+            if (meta) {
+                *meta = iter;
+            }
+            RBT_GET(iter, rbt);
+            return rbt;
+        }
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Call plugins_types sort callback or sort values by plugin order.
+ *
+ * @param[in] ctx libyang context.
+ * @param[in] val1 First value to compare.
+ * @param[in] val2 Second value to compare.
+ * @return Negative number if val1 < val2,
+ * @return Zero if val1 == val2,
+ * @return Positive number if val1 > val2.
+ */
+static int
+rb_sort_clb(const struct ly_ctx *ctx, const struct lyd_value *val1, const struct lyd_value *val2)
+{
+    assert(val1->realtype == val2->realtype);
+    return val1->realtype->plugin->sort(ctx, val1, val2);
+}
+
+/**
+ * @brief Compare red-black nodes by rb_node.dnode from the same Red-black tree.
+ *
+ * @param[in] n1 First leaf-list data node.
+ * @param[in] n2 Second leaf-list data node.
+ * @return Negative number if val1 < val2,
+ * @return Zero if val1 == val2,
+ * @return Positive number if val1 > val2.
+ */
+static int
+rb_compare_leaflists(const struct lyd_node *n1, const struct lyd_node *n2)
+{
+    struct lyd_value *val1, *val2;
+
+    /* compare leaf-list values */
+    assert(n2->schema->nodetype == LYS_LEAFLIST);
+    assert(n1->schema->nodetype == LYS_LEAFLIST);
+
+    val1 = &((struct lyd_node_term *)n1)->value;
+    val2 = &((struct lyd_node_term *)n2)->value;
+    return rb_sort_clb(LYD_CTX(n1), val1, val2);
+}
+
+/**
+ * @brief Compare red-black nodes by rb_node.dnode from the same Red-black tree.
+ *
+ * @param[in] n1 First list data node.
+ * @param[in] n2 Second list data node.
+ * @return Negative number if val1 < val2,
+ * @return Zero if val1 == val2,
+ * @return Positive number if val1 > val2.
+ */
+static int
+rb_compare_lists(const struct lyd_node *n1, const struct lyd_node *n2)
+{
+    const struct lyd_node *k1, *k2;
+    struct lyd_value *val1, *val2;
+    int cmp;
+
+    /* compare first list key */
+    assert(n1->schema->nodetype & LYS_LIST);
+    assert(n2->schema->nodetype & LYS_LIST);
+    cmp = 0;
+    k1 = ((const struct lyd_node_inner *)n1)->child;
+    k2 = ((const struct lyd_node_inner *)n2)->child;
```

Comment:
Instead use `lyd_child(n1)`.

Reply:
The `perf` tool complained that `lyd_child()` was using too much CPU.